File: src/octave.cpp
#include <vector>
#include <cmath>
#include <iostream>
#include <emscripten/bind.h>

#include "octave.hpp"

Octave::Octave(size_t audio_size, size_t num_bins, unsigned int sample_rate_)
  : sample_rate(sample_rate_) {

  // Allocate storage devices
  slices   = std::vector<float>(num_bins);
  window   = ComplexArray(audio_size);
  window_d = ComplexArray(audio_size);
  hann     = std::vector<float>(audio_size);
  hann_d   = std::vector<float>(audio_size);

  // Compute the Hann window
  for (size_t i = 0; i < audio_size; i++) {
    // center of the window is n = 0
    float N = audio_size;
    float n = i - (N - 1)/2.;
    hann[i] = 0.5 * (1 + std::cos(2 * M_PI * n/(N - 1)));
    hann_d[i] =  -(M_PI * sample_rate)/(N - 1) * std::sin(2 * M_PI * n/(N - 1));
  }
}

std::vector<float> & Octave::audio_to_slices(const std::vector<float> & audio) {
  // Reset the slices
  for (size_t i = 0; i < slices.size(); i++) {
    slices[i] = 0;
  }

  // Window the audio
  for (size_t i = 0; i < window.size(); i++) {
    window[i]   = hann[i]   * audio[i];
    window_d[i] = hann_d[i] * audio[i];
  }

  // Compute the FFT
  fft(window);
  fft(window_d);

  for (size_t i = 0; i < window.size()/2 + 1; i++) {
    // Compute the frequency reassignment
    float freq = (2 * M_PI * i * sample_rate)/float(window.size());
    float dphase_dt = -std::imag(window_d[i] * std::conj(window[i])/std::norm(window[i]));
    float freq_reassigned = freq + dphase_dt;
    //if (freq_reassigned <= 0) continue;

    if (freq_reassigned < 2 * M_PI * 60 or freq_reassigned > 2 * M_PI * 3000) continue;

    // Wrap it in the octave
    float wrapped_freq = std::fmod(std::log2(freq_reassigned), 1);
    while (wrapped_freq < 0) wrapped_freq += 1;

    // Find the nearest bin
    float approx_bin = wrapped_freq * slices.size();
    int left_bin = std::floor(approx_bin);
    float right = approx_bin - left_bin;
    float value = std::abs(window[i]);
    slices[left_bin % slices.size()] += (1 - right) * value;
    slices[(left_bin + 1) % slices.size()] += right * value;
  }

  return slices;
}

void Octave::fft(ComplexArray & x) {
  // Source: https://tfetimes.com/c-fast-fourier-transform/
  const size_t N = x.size();
  if (N <= 1) return;

  // Divide
  ComplexArray even = x[std::slice(0, N/2, 2)];
  ComplexArray  odd = x[std::slice(1, N/2, 2)];

  // conquer
  fft(even);
  fft(odd);

  // combine
  for (size_t k = 0; k < N/2; k++) {
    Complex t = std::polar(1.0, -2 * M_PI * k / N) * odd[k];
    x[k    ] = even[k] + t;
    x[k+N/2] = even[k] - t;
  }
}

// Binding code
using namespace emscripten;
EMSCRIPTEN_BINDINGS(octave) {
  class_<Octave>("Octave")
  .constructor<size_t, size_t, unsigned int>()
  .function("audioToSlices", &Octave::audio_to_slices)
  ;
  register_vector<float>("VectorFloat");
}

File: js/octave.js
function Octave(audioSize, numBins, sampleRate) {
  this.sampleRate = sampleRate;

  // Allocate storage
  this.slices   = new Float32Array(numBins);
  this.windowR  = new Float32Array(audioSize);
  this.windowI  = new Float32Array(audioSize);
  this.windowDR = new Float32Array(audioSize);
  this.windowDI = new Float32Array(audioSize);
  this.audio    = new Float32Array(audioSize);
  this.hann     = new Float32Array(audioSize);
  this.hannD    = new Float32Array(audioSize);

  // Precompute the Hann windows
  for (var i = 0; i < audioSize; i++) { 
    var N = audioSize;
    var n = i - (N - 1)/2;
    var angle = (2 * Math.PI * n)/(N - 1);
    this.hann[i] = 0.5 * (1 + Math.cos(angle));
    this.hannD[i] = -(Math.PI * this.sampleRate)/(N - 1) * Math.sin(angle);
  }
}

Octave.prototype.processAudio = function() {
  // Reset the slices
  this.slices.fill(0);

  // Window the audio
  for (var i = 0; i < this.windowR.length; i++) {
    this.windowR [i] = this.hann [i] * this.audio[i];
    this.windowDR[i] = this.hannD[i] * this.audio[i];
  }
  this.windowI.fill(0);
  this.windowDI.fill(0);
  
  // Compute the FFT
  FFT(this.windowR, this.windowI);
  FFT(this.windowDR, this.windowDI);

  // Wrap it to the octave
  for (var i = 1; i < this.windowR.length/2 + 1; i++) {
    // Precompute the norm
    var norm = this.windowR[i] * this.windowR[i] + this.windowI[i] * this.windowI[i]; 

    // Compute the frequency
    var freq = (2 * Math.PI * i * this.sampleRate)/(this.windowR.length);
    var dPhaseDT = (this.windowI[i] * this.windowDR[i] - this.windowR[i] * this.windowDI[i])/norm;
    var freqReassigned = freq + dPhaseDT;
    var wrappedFreq = Math.log2(freqReassigned) % 1;

    // Find the bin
    this.placeSlice(wrappedFreq, Math.sqrt(norm));
  }
}

Octave.prototype.placeSlice = function(position, value) {
  var bin = position * this.slices.length;
  var leftOfBin = Math.floor(bin);
  var rightPercent = bin - leftOfBin;
  this.slices[leftOfBin % this.slices.length] += (1 - rightPercent) * value;
  this.slices[(leftOfBin + 1) % this.slices.length] += rightPercent * value;
}

function FFT(real, imag) {
 // https://www.nayuki.io/res/free-small-fft-in-multiple-languages/fft.js
 // Length variables
 var n = real.length;
 if (n != imag.length)
  throw "Mismatched lengths";
 if (n == 1)  // Trivial transform
  return;
 var levels = -1;
 for (var i = 0; i < 32; i++) {
  if (1 << i == n)
   levels = i;  // Equal to log2(n)
 }
 if (levels == -1)
  throw "Length is not a power of 2";

 // Trigonometric tables
 var cosTable = new Array(n / 2);
 var sinTable = new Array(n / 2);
 for (var i = 0; i < n / 2; i++) {
  cosTable[i] = Math.cos(2 * Math.PI * i / n);
  sinTable[i] = Math.sin(2 * Math.PI * i / n);
 }

 // Bit-reversed addressing permutation
 for (var i = 0; i < n; i++) {
  var j = reverseBits(i, levels);
  if (j > i) {
   var temp = real[i];
   real[i] = real[j];
   real[j] = temp;
   temp = imag[i];
   imag[i] = imag[j];
   imag[j] = temp;
  }
 }

 // Cooley-Tukey decimation-in-time radix-2 FFT
 for (var size = 2; size <= n; size *= 2) {
  var halfsize = size / 2;
  var tablestep = n / size;
  for (var i = 0; i < n; i += size) {
   for (var j = i, k = 0; j < i + halfsize; j++, k += tablestep) {
    var l = j + halfsize;
    var tpre =  real[l] * cosTable[k] + imag[l] * sinTable[k];
    var tpim = -real[l] * sinTable[k] + imag[l] * cosTable[k];
    real[l] = real[j] - tpre;
    imag[l] = imag[j] - tpim;
    real[j] += tpre;
    imag[j] += tpim;
   }
  }
 }

 // Returns the integer whose value is the reverse of the lowest 'width' bits of the integer 'val'.
 function reverseBits(val, width) {
  var result = 0;
  for (var i = 0; i < width; i++) {
   result = (result << 1) | (val & 1);
   val >>>= 1;
  }
  return result;
 }
}
